Import YAML file.

DavidACosgrove opened the pull request:
Added YAML for arbitrary reactions.

jones-gareth commented:
I think this needs a better name that "Arbitrary reactions".  Perhaps "Reaction runner" or "Reaction transform"

jones-gareth commented:
As in the PAINS function I'm not seeing the bond highlighting.  I will debug that.

jones-gareth commented:
I got this error while trying to run the data function: 

```
Python data function error

Click to see details of error output and any traceback

If reporting an error on a supported data function, copy this notification text

[14:23:29] Explicit valence for atom # 4 C, 5, is greater than permitted
Traceback (most recent call last):
  File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\main.py", line 18, in <module>
    response = execute(request)
  File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 85, in execute
    products = run_reactions(mols, rxn)
  File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 51, in run_reactions
    Chem.SanitizeMol(new_prod)
rdkit.Chem.rdchem.AtomValenceException: Explicit valence for atom # 4 C, 5, is greater than permitted
```

These sanitization issues are, unfortunately, quite common.  There is a function `ruse.rdkit.rdkit_utils.sanitize_mol` that attempts to handle these issues.

DavidACosgrove commented:
I guess I need to sanitize explicitly in the function and catch any
exceptions.  I’ll sort it tomorrow morning.

On Mon, 5 Sep 2022 at 21:28, Gareth Jones ***@***.***> wrote:

> I got this error while trying to run the data function:
>
> Python data function error
>
> Click to see details of error output and any traceback
>
> If reporting an error on a supported data function, copy this notification text
>
> [14:23:29] Explicit valence for atom # 4 C, 5, is greater than permitted
> Traceback (most recent call last):
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\main.py", line 18, in <module>
>     response = execute(request)
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 85, in execute
>     products = run_reactions(mols, rxn)
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 51, in run_reactions
>     Chem.SanitizeMol(new_prod)
> rdkit.Chem.rdchem.AtomValenceException: Explicit valence for atom # 4 C, 5, is greater than permitted
>
> These sanitization issues are, unfortunately, quite common. There is a
> function ruse.rdkit.rdkit_utils.sanitize_mol that attempts to handle
> these issues.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/3#issuecomment-1237440657>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FX6GPIQVMMA3FW42YTV4ZJWFANCNFSM6AAAAAAQEKYBPE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
Should we highlight the transformed part of the molecule?

DavidACosgrove commented:
I've put the sanitize_mol call in, but I could do with the reaction you
used that gave this error so I can put a test in for it.

On Mon, Sep 5, 2022 at 9:28 PM Gareth Jones ***@***.***>
wrote:

> I got this error while trying to run the data function:
>
> Python data function error
>
> Click to see details of error output and any traceback
>
> If reporting an error on a supported data function, copy this notification text
>
> [14:23:29] Explicit valence for atom # 4 C, 5, is greater than permitted
> Traceback (most recent call last):
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\main.py", line 18, in <module>
>     response = execute(request)
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 85, in execute
>     products = run_reactions(mols, rxn)
>   File "C:\Users\jones\AppData\Local\Temp\ChemChartsDataFxns\jobs\8983ddf9-2a94-44a7-bebe-53ecbbdb85a7\script.py", line 51, in run_reactions
>     Chem.SanitizeMol(new_prod)
> rdkit.Chem.rdchem.AtomValenceException: Explicit valence for atom # 4 C, 5, is greater than permitted
>
> These sanitization issues are, unfortunately, quite common. There is a
> function ruse.rdkit.rdkit_utils.sanitize_mol that attempts to handle
> these issues.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/3#issuecomment-1237440657>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FX6GPIQVMMA3FW42YTV4ZJWFANCNFSM6AAAAAAQEKYBPE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
> I've put the sanitize_mol call in, but I could do with the reaction you used that gave this error so I can put a test in for it.

The reaction I drew was this:

![image](https://user-images.githubusercontent.com/24233741/188750756-229baec8-e681-4657-b1ec-5b7fac68156a.png)

With the sanitize_mol call it no longer crashes, but I get structures like this:

![image](https://user-images.githubusercontent.com/24233741/188751079-abfb3b84-5490-4876-abd3-9eb0b89370aa.png)

The original valence error is from the hydroxy being added twice to a carbon atom- which looks like a bug.

Highlighting is working.

jones-gareth commented on python/local/ArbitraryReaction.yaml:
```diff
@@ -0,0 +1,165 @@
+    contentType: chemical/x-mdl-rxnfile
+tags:
+  - color: '#50AF28'
+    text: chemistry
```
Remove calcprop tag

DavidACosgrove commented:
Before I get to the bug, this does highlight the fact that my decision to
run all possible reactions on the molecule to give a single product is
probably wrong, though I'm not sure what else to do.  I'm fairly confident
the user would be disappointed by receiving the penta-hydroxy product in
this case though any mono-hydroxy result is clearly going to be arbitrary.
Some discussion and a better decision about how to handle this would be
useful.

On Tue, Sep 6, 2022 at 11:36 PM Gareth Jones ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In python/local/ArbitraryReaction.yaml
> <https://github.com/Glysade/DataFxns/pull/3#discussion_r964231271>:
>
> > +    type: molecule
> +    sketcherContentType: chemical/x-mdl-rxnfile
> +    molTypes:
> +    - extension: .sma
> +      contentType: chemical/x-smarts
> +      name: SMARTS (*.sma)
> +    - extension: .rxn
> +      contentType: chemical/x-mdl-rxnfile
> +      name: RXN file (*.rxn)
> +  request:
> +    id: reactionQuery
> +    dataType: string
> +    contentType: chemical/x-mdl-rxnfile
> +tags:
> +  - color: '#50AF28'
> +    text: chemistry
>
> Remove calcprop tag
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/3#pullrequestreview-1098326359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FSUWTOD7C7PS2FVQ3DV47BQFANCNFSM6AAAAAAQEKYBPE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

DavidACosgrove commented:
The bad molecule is caused by the fact that sanitize_mol allows
AtomValenceException and KekulizeException.  I've changed my code to use
SanitizeMol directly and catch any MolSanitizeException and discard those
molecules.  It now behaves as expected except that putting an OH on the
fusion atoms doesn't raise an exception so is allowed.  That does, however,
produce a SMILES that the SMILES parser doesn't accept so I've added that
as a 2nd test.  I will put an issue into the RDKit github to see what
people think, because it feels like a bug to me.

On Tue, Sep 6, 2022 at 11:33 PM Gareth Jones ***@***.***>
wrote:

> I've put the sanitize_mol call in, but I could do with the reaction you
> used that gave this error so I can put a test in for it.
>
> The reaction I drew was this:
>
> [image: image]
> <https://user-images.githubusercontent.com/24233741/188750756-229baec8-e681-4657-b1ec-5b7fac68156a.png>
>
> With the sanitize_mol call it no longer crashes, but I get structures like
> this:
>
> [image: image]
> <https://user-images.githubusercontent.com/24233741/188751079-abfb3b84-5490-4876-abd3-9eb0b89370aa.png>
>
> The original valence error is from the hydroxy being added twice to a
> carbon atom- which looks like a bug.
>
> Highlighting is working.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/3#issuecomment-1238718694>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUDF7AW5RM2RTTN633V47BDFANCNFSM6AAAAAAQEKYBPE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
> Before I get to the bug, this does highlight the fact that my decision to run all possible reactions on the molecule to give a single product is probably wrong, though I'm not sure what else to do. I'm fairly confident the user would be disappointed by receiving the penta-hydroxy product in this case though any mono-hydroxy result is clearly going to be arbitrary. Some discussion and a better decision about how to handle this would be useful.
> […](#)
> On Tue, Sep 6, 2022 at 11:36 PM Gareth Jones ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In python/local/ArbitraryReaction.yaml <[#3 (comment)](https://github.com/Glysade/DataFxns/pull/3#discussion_r964231271)>: > + type: molecule + sketcherContentType: chemical/x-mdl-rxnfile + molTypes: + - extension: .sma + contentType: chemical/x-smarts + name: SMARTS (*.sma) + - extension: .rxn + contentType: chemical/x-mdl-rxnfile + name: RXN file (*.rxn) + request: + id: reactionQuery + dataType: string + contentType: chemical/x-mdl-rxnfile +tags: + - color: '#50AF28' + text: chemistry Remove calcprop tag — Reply to this email directly, view it on GitHub <[#3 (review)](https://github.com/Glysade/DataFxns/pull/3#pullrequestreview-1098326359)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACGF2FSUWTOD7C7PS2FVQ3DV47BQFANCNFSM6AAAAAAQEKYBPE> . You are receiving this because you authored the thread.Message ID: ***@***.***>
> -- David Cosgrove Freelance computational chemistry and chemoinformatics developer http://cozchemix.co.uk

It's a tricky one- I can imagine in some cases that the user might want to apply the transform to all matches (deprotection for example), whereas there are other times that they'll only want the transform applied once.  I think that there should be an option for this- though I'll let Matt and Paul weigh in.
As you say, if the transform has multiple matches, but the user only wants a single application then the reaction site will be arbitrary.  I can't see what we can do about that.

mattstahl00 commented:
There could be 3 behavior's produced by the DataFxn -- apply the rxn once (1 output struct per row), apply all rxns (1 output struct per row), and apply all rxns with output to a new data table with 1 row per product.  The output type would be determined by the user selections.  The first two options would add columns to existing tables.  The last one would create a new data table as there would be more rows than the input table.

jones-gareth commented:
Closed as it is redundant.
